### PR TITLE
feat(server): implement JSON.TOGGLE & JSON.ARRLEN & JSON.OBJLEN commands

### PR DESCRIPTION
**Pull request by iko1**
Signed-off-by: iko1 <me@remotecpp.dev>

fixes: #241

### CODE REVIEW

File: src/server/json_family.cc [hunk 1]
```diff
@@ -70,6 +73,41 @@ string JsonType(const json& val) {
   return "";
 }
 
+template<class Json>
+error_code json_replace_with_ec(Json& instance, string_view& path, JsonReplaceCb callback,
```
Comment (romange) [inline src/server/json_family.cc:77]: the jsoncons library uses templated classes for maximum flexibility. Here, I suspect you do not need it since we always use `json` type. Why not get rid of the template? Also, the naming convention for function names is `CamelCase`.

File: src/server/json_family.cc [hunk 2]
```diff
@@ -164,8 +202,159 @@ OpResult<vector<OptSizeT>> OpStrLen(const OpArgs& op_args, string_view key, Json
   return vec;
 }
 
+OpResult<vector<OptSizeT>> OpObjLen(const OpArgs& op_args, string_view key, JsonExpression expression) {
+  OpResult<PrimeIterator> it_res = op_args.shard->db_slice().Find(op_args.db_ind, key, OBJ_STRING);
```
Comment (romange) [inline src/server/json_family.cc:206]: it seems that lines 206-222 repeat themselves several times in this file. If I am not mistaken, then it's better to refactor them to a utility function that returns `OpResult<json>`

File: src/server/json_family.cc [hunk 3]
```diff
@@ -164,8 +202,159 @@ OpResult<vector<OptSizeT>> OpStrLen(const OpArgs& op_args, string_view key, Json
...
+      for (auto& it: *result) {
+        if (it.has_value()) {
+          (*cntx)->SendLong(*it);
+        } else {
+          (*cntx)->SendNull();
+        }
+      }
+    }
+  } else {
+    switch (result.status()) {
```
Comment (romange) [inline src/server/json_family.cc:344]: I think if you just call `(*cntx)->SendError(result.status())` you will get a behavior equivalent to that switch.   
if not, please tell me where it breaks.

File: src/server/json_family_test.cc [hunk 4]
```diff
@@ -50,7 +50,7 @@ TEST_F(JsonFamilyTest, SetGetBasic) {
     </store>
 )";
 
-  auto resp = Run({"set", "json", json});
+  auto resp = Run({"SET", "json", json});
```
Comment (romange) [inline src/server/json_family_test.cc:53]: why did you change the case? it should work regardless.

File: src/server/json_family_test.cc [hunk 5]
```diff
@@ -155,8 +155,96 @@ TEST_F(JsonFamilyTest, StrLen) {
...
+    {"a":true, "b":false, "c":1, "d":null, "e":"foo", "f":[], "g":{}}
+  )";
+
+  auto resp = Run({"SET", "json", json});
+  ASSERT_THAT(resp, "OK");
+
+  resp = Run({"JSON.TOGGLE", "json", "$.*"});
+  ASSERT_THAT(resp, ArgType(RespExpr::ARRAY));
+  EXPECT_THAT(resp.GetVec(), ElementsAre(IntArg(0), IntArg(1), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL)));
+
```
Comment (romange) [inline src/server/json_family_test.cc:244]: you validate that response is correct but you do not check that the actual json object has been changed.
Please read some of the fields and validate that their values are updated.

**Review by romange — CHANGES_REQUESTED** [review]
Thanks for implementing this!

File: src/server/json_family_test.cc (hunk 4 above)
Comment (iko1) [reply src/server/json_family_test.cc:53]: I tried to replace a specific word but I clicked on 'replace all' by mistake.
I'm going to revert this change

File: src/server/json_family_test.cc (hunk 5 above)
Comment (iko1) [reply src/server/json_family_test.cc:244]: Which command should I use to retrieve the actual JSON?
1. GET
2`JSON.GET json $.*` that returns the output embedded within Json array  e.g.:
`'{"a":true, "b":false, "c":1, "d":null, "e":"foo", "f":[], "g":{}}'d`"[true,false,1,null,\"foo\",[],{}]"`
3. Refactor `JSON.GET` to return the actual Json if the path isn't specified e.g. `JSON.GET json`

Comment (romange) [reply src/server/json_family_test.cc:244]: I think `json.get json $.*` is great here.

File: src/server/json_family_test.cc [hunk 6]
```diff
@@ -136,9 +136,6 @@ TEST_F(JsonFamilyTest, Type) {
 
   resp = Run({"JSON.TYPE", "json", "$[10]"});
   EXPECT_THAT(resp, ArgType(RespExpr::NIL));
-
-  resp = Run({"JSON.TYPE", "not_exist_key", "$[10]"});
-  EXPECT_THAT(resp, ArgType(RespExpr::NIL_ARRAY));
```
Comment (romange) [inline src/server/json_family_test.cc:141]: what happens now with a non-existent key?

File: src/server/json_family_test.cc [hunk 7]
```diff
@@ -155,8 +152,280 @@ TEST_F(JsonFamilyTest, StrLen) {
...
+
+  auto resp = Run({"set", "json", json});
+  ASSERT_THAT(resp, "OK");
+
+  resp = Run({"JSON.TOGGLE", "json", "$.*"});
+  ASSERT_THAT(resp, ArgType(RespExpr::ARRAY));
+  EXPECT_THAT(resp.GetVec(), ElementsAre(IntArg(0), IntArg(1), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL), ArgType(RespExpr::NIL)));
+
+  resp = Run({"JSON.GET", "json", "$.*"});
+  EXPECT_EQ(resp, string_view(R"([false,true,1,null,"foo",[],{}])", 31));
```
Comment (romange) [inline src/server/json_family_test.cc:243]: why did you need to wrap it with explicit string_view c'tor?

File: src/server/json_family_test.cc (hunk 6 above)
Comment (iko1) [reply src/server/json_family_test.cc:141]: It returns an error "no such key", instead of what is written in Elasticache spec to return a NULL array.
I think returning an error is more indicative in this case.

File: src/server/json_family_test.cc (hunk 7 above)
Comment (iko1) [reply src/server/json_family_test.cc]: Thanks for this comment. It's definitely redundant, so I removed those explicit string_view c'tors.

File: src/server/json_family.cc [hunk 8]
```diff
@@ -246,16 +253,144 @@ OpResult<vector<OptBool>> OpToggle(const OpArgs& op_args, string_view key, strin
...
+
+  if (result) {
+    DVLOG(1) << "JSON.NUMINCRBY " << trans->DebugId() << ": " << key;
+    if (result->empty()) {
+      (*cntx)->SendNullArray();
+    } else {
+      (*cntx)->StartArray(result->size());
+      for (auto& it: *result) {
+        if (it.has_value()) {
+          (*cntx)->SendLong(*it);
```
Comment (iko1) [inline src/server/json_family.cc:345]: @romange, Should I implement a function that retrieves a `long long` value, right?

File: src/server/json_family.cc [hunk 9]
```diff
@@ -89,83 +134,294 @@ OpResult<string> OpGet(const OpArgs& op_args, string_view key, vector<pair<strin
...
+OpResult<vector<OptLongLong>> OpNumMultBy(const OpArgs& op_args, string_view key, string_view path, long long num) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptLongLong> vec;
+  auto cb = [&](const string& path, json& val) {
+    if (val.is_number()) {
+      auto current_val = val.as<long long>() * num;
```
Comment (iko1) [inline src/server/json_family.cc:297]: I should return an OVERFLOW error if the result is out of the range of 64-bit IEEE double.

File: src/server/json_family_test.cc (hunk 6 above)
Comment (romange) [reply src/server/json_family_test.cc:141]: It's not up to us to change the spec at this point. I just whitelist these special cases explicitly.
See https://github.com/dragonflydb/dragonfly/blob/main/src/server/zset_family.cc#L1541 for example

File: src/server/json_family.cc (hunk 9 above)
Comment (romange) [reply src/server/json_family.cc:297]: In order to provide fully compatible behavior, we need to setup an elasticache instance and learn how these special cases are handled there. That's out of scope of this assignment at this point.

what I usually do is I write `// TODO: .... ` comments when the code needs further enhancement.

File: src/server/json_family.cc (hunk 8 above)
Comment (romange) [reply src/server/json_family.cc:345]: you mean like `SendLong`? Then no, because on 64bit architectures `long long` and `long` is the same and we are only 64bit.

File: src/server/json_family.cc [hunk 10]
```diff
@@ -89,83 +134,294 @@ OpResult<string> OpGet(const OpArgs& op_args, string_view key, vector<pair<strin
...
+  return vec;
+}
+
+OpResult<vector<OptLongLong>> OpNumIncrBy(const OpArgs& op_args, string_view key, string_view path, long long num) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptLongLong> vec;
```
Comment (romange) [inline src/server/json_family.cc:266]: please add a comment `TODO: to check how double and boolean are handled by INCRBY`

**Comment by romange** [comment]
Please see #240. I suggest that you copy `json.NUMINCRBY` and `json.NUMMULTBY` changes aside and pull them out of this PR. For several reasons:
1. the title of this PR corresponds to other commands.
2. they need a significant rewrite because internally you need to manage doubles and not long long and this PR becomes too big anyway.

**Comment by iko1** [comment]
> Please see #240. I suggest that you copy `json.NUMINCRBY` and `json.NUMMULTBY` changes aside and pull them out of this PR. For several reasons:
> 
>     1. the title of this PR corresponds to other commands.
> 
>     2. they need a significant rewrite because internally you need to manage doubles and not long long and this PR becomes too big anyway.

This PR contains only the commands described in the title. Other JSON commands that were related to this PR will copy to another PR with a significant refactor(Thanks for the comments).

File: src/server/json_family.cc [hunk 11]
```diff
@@ -70,7 +83,38 @@ string JsonType(const json& val) {
...
+    jsoncons::jsonpath::detail::static_resources<value_type,reference> static_resources(funcs);
+    evaluator_t e;
+    json_selector_t expr = e.compile(static_resources, path, ec);
+    if (ec) {
+      return ec;
+    }
+
+    jsoncons::jsonpath::detail::dynamic_resources<value_type,reference> resources;
+    auto f = [&callback](const json_location_type& path, reference val)
+    {
```
Comment (romange) [inline src/server/json_family.cc:104]: i think this file is not formatted according to .clang-format

File: src/server/json_family.cc [hunk 12]
```diff
@@ -70,7 +83,38 @@ string JsonType(const json& val) {
   return "";
 }
 
-OpResult<string> OpGet(const OpArgs& op_args, string_view key, vector<pair<string_view, JsonExpression>> expressions) {
+error_code JsonReplace(json& instance, string_view& path, JsonReplaceCb callback) {
+    using evaluator_t = typename jsoncons::jsonpath::detail::jsonpath_evaluator<json, json&>;
```
Comment (romange) [inline src/server/json_family.cc:87]: nit: you do not need "typename" anymore

Comment (iko1) [reply src/server/json_family.cc]: It has been removed now

File: src/server/json_family_test.cc (hunk 6 above)
Comment (romange) [reply src/server/json_family_test.cc:141]: this still requires a fix :)

File: src/server/json_family.cc [hunk 13]
```diff
@@ -89,82 +134,163 @@ OpResult<string> OpGet(const OpArgs& op_args, string_view key, vector<pair<strin
...
+  DVLOG(1) << "Before Get::ScheduleSingleHopT " << key;
+  Transaction* trans = cntx->transaction;
+  OpResult<vector<OptBool>> result = trans->ScheduleSingleHopT(move(cb));
+
+  if (result) {
+    DVLOG(1) << "JSON.TOGGLE " << trans->DebugId() << ": " << key;
+    if (result->empty()) {
+      (*cntx)->SendNullArray();
+    } else {
+      (*cntx)->StartArray(result->size());
```
Comment (romange) [inline src/server/json_family.cc:281]: there are three repetitions of the same code of sending a long array.
We can factor them out in the next changes.

File: src/server/json_family_test.cc [hunk 14]
```diff
@@ -107,19 +107,19 @@ TEST_F(JsonFamilyTest, SetGetFromPhonebook) {
...
-  EXPECT_EQ(resp, string_view("[]", 2));
+  EXPECT_EQ(resp, "[]");
 
   resp = Run({"JSON.GET", "json", "$.children.*"});
-  EXPECT_EQ(resp, string_view("[]", 2));
+  EXPECT_EQ(resp, "[]");
 
   resp = Run({"JSON.GET", "json", "$..phoneNumbers[1].*"});
-  EXPECT_EQ(resp, string_view(R"(["646 555-4567","office"])", 25));
+  EXPECT_EQ(resp, R"(["646 555-4567","office"])");
```
Comment (romange) [inline src/server/json_family_test.cc:122]: what happens if you call `Run({"JSON.GET", "json", "$.none"}); ` ?
the correct behavior is to return an empty array.

File: src/server/json_family.cc (hunk 13 above)
Comment (iko1) [reply src/server/json_family.cc:281]: I put it as a requirement for the next PR.

**Review by romange — APPROVED** [review]
